grype/db/v6/blob_store.go

```diff
@@ -0,0 +1,112 @@
+package v6
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"gorm.io/gorm"
+
+	"github.com/anchore/grype/internal/log"
+)
+
+type blobable interface {
+	getBlobValue() any
+	setBlobID(int64)
+}
+
+type blobStore struct {
+	db *gorm.DB
+}
+
+func newBlobStore(db *gorm.DB) *blobStore {
+	return &blobStore{
+		db: db,
+	}
+}
+
+func (s *blobStore) addBlobable(bs ...blobable) error {
+	for i := range bs {
+		b := bs[i]
+		v := b.getBlobValue()
+		if v == nil {
+			continue
+		}
+		bl := newBlob(v)
+
+		if err := s.addBlobs(bl); err != nil {
+			return err
+		}
+
+		b.setBlobID(bl.ID)
+	}
+	return nil
+}
+
+func (s *blobStore) addBlobs(blobs ...*Blob) error {
+	for i := range blobs {
+		v := blobs[i]
+		digest := v.computeDigest()
+
+		var blobDigest BlobDigest
+		err := s.db.Where("id = ?", digest).First(&blobDigest).Error
+		if err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
+			return fmt.Errorf("failed to get blob digest: %w", err)
+		}
+
+		if blobDigest.BlobID != 0 {
+			v.ID = blobDigest.BlobID
+			continue
+		}
+
+		if err := s.db.Create(v).Error; err != nil {
+			return fmt.Errorf("failed to create blob: %w", err)
+		}
+
+		blobDigest = BlobDigest{
+			ID:     digest,
+			BlobID: v.ID,
+		}
+		if err := s.db.Create(blobDigest).Error; err != nil {
+			return fmt.Errorf("failed to create blob digest: %w", err)
+		}
+	}
+	return nil
+}
+
+func (s *blobStore) getBlobValue(id int64) (string, error) {
+	var blob Blob
+	if err := s.db.First(&blob, id).Error; err != nil {
+		return "", err
+	}
+	return blob.Value, nil
+}
+
+func (s *blobStore) Close() error {
+	var count int64
+	if err := s.db.Model(&Blob{}).Count(&count).Error; err != nil {
```

Comment:
I don't really want to overrotate on performance concerns so don't want to refactor this to avoid writing temp data (dropping tables is cheap, and writing the digests is also relatively cheap, and the unique constraint is useful on writes). This Close() method is called by the parent store Close() only when we have enabled writes to the DB, so this is a nop when used as a reader. I've added a comment to explain some of this.